src/utils/display-observer.js

```diff
@@ -117,24 +170,94 @@ export class DisplayObserver {
 
   /** @override */
   dispose() {
-    this.observers_.forEach((observer) => observer.disconnect());
+    this.observers_.forEach((observer) => {
+      if (observer.io) {
+        observer.io.disconnect();
+      }
+    });
+    this.observers_.length = 0;
     this.visibilityUnlisten_();
     this.visibilityUnlisten_ = null;
     this.targetObserverCallbacks_.clear();
     this.targetObservations_.clear();
   }
 
+  /**
+   * @param {!Element} container
+   * @param {?Element=} opt_root
+   */
+  registerContainer(container, opt_root) {
+    const index = findObserverByContainer(this.observers_, container);
+    if (index != -1) {
+      return;
+    }
+
+    this.observers_.push({
+      container,
+      root: opt_root || container,
+      contains: (target) => containsNotSelf(container, target),
+      // Start with null as IntersectionObserver. Will be initialized when
+      // the container itself becomes displayed.
+      io: null,
+    });
+    this.observe(container, this.containerObserved_);
+  }
+
+  /**
+   * @param {!Element} container
+   */
+  unregisterContainer(container) {
+    const index = findObserverByContainer(this.observers_, container);
+    if (index < CUSTOM_CONTAINER_OFFSET) {
+      // The container has been unregistered already.
+      return;
+    }
+
+    // Remove observer.
+    const observer = this.observers_[index];
+    this.observers_.splice(index, 1);
+    if (observer.io) {
+      observer.io.disconnect();
+    }
+
+    // Unobserve the container itself.
+    this.unobserve(container, this.containerObserved_);
+
+    // Remove observations.
+    this.targetObserverCallbacks_.forEach((callbacks, target) => {
+      const observations = this.targetObservations_.get(target);
+      if (!observations || observations.length <= index) {
+        return;
+      }
+      const oldDisplay = computeDisplay(observations, this.isDocDisplay_);
+      observations.splice(index, 1);
+      const newDisplay = computeDisplay(observations, this.isDocDisplay_);
+      notifyIfChanged(callbacks, target, newDisplay, oldDisplay);
+    });
+  }
+
   /**
    * @param {!Element} target
-   * @param {function(boolean)} callback
+   * @param {!ObserverCallbackDef} callback
    */
   observe(target, callback) {
     let callbacks = this.targetObserverCallbacks_.get(target);
     if (!callbacks) {
       callbacks = [];
       this.targetObserverCallbacks_.set(target, callbacks);
+
+      // Subscribe observers.
       for (let i = 0; i < this.observers_.length; i++) {
-        this.observers_[i].observe(target);
+        const observer = this.observers_[i];
+        if (observer.io && observer.contains(target)) {
```

Comment:
I need to check on `observer.io` here because later I'd have to use it to actually observe the target. So if it's not there, it means that the target can be considered to be not-displayed on it.